### PR TITLE
[Merged by Bors] - Implement `Object.prototype.__defineGetter__`

### PR DESCRIPTION
**Pull request by CYBAI**
This Pull Request fixes part of #2067.

It introduces the legacy `Object.prototype.__defineGetter__` function.

### CODE REVIEW

**Comment by Razican** [comment]
### Test262 conformance changes

  #### VM implementation
  | Test result | main count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 90,499 | 90,499 | 0 |
  | Passed | 56,372 | 56,384 | **+12** |
  | Ignored | 23,580 | 23,580 | 0 |
  | Failed | 10,547 | 10,535 | **-12** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 62.29% | 62.30% | **+0.01%** |
  <details><summary><b>Fixed tests (12):</b></summary>

  ```
  test/built-ins/Object/prototype/__defineGetter__/this-non-obj.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/this-non-obj.js (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/length.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/length.js (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/name.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/name.js (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/prop-desc.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/prop-desc.js (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/getter-non-callable.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/getter-non-callable.js (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/define-non-configurable.js [strict mode] (previously Failed)
  test/built-ins/Object/prototype/__defineGetter__/define-non-configurable.js (previously Failed)
  ```
  </details>

File: boa_engine/src/builtins/object/mod.rs [hunk 1]
```diff
@@ -115,6 +116,41 @@ impl Object {
...
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        let getter = args.get_or_undefined(1);
+
+        let obj = this.to_object(context)?;
+
+        this.as_callable().ok_or_else(|| {
+            context.construct_type_error("Object.prototype.__defineGetter__: Expecting function")
+        })?;
```
Comment (raskad) [inline boa_engine/src/builtins/object/mod.rs:140]: This should test if `getter` is callable not `this`. From the spec: `If IsCallable(getter) is false`.

Also we can avoid the closure by using `getter.is_callable()` and then `return context.throw_type_error()` to throw the error.

**Review by raskad — CHANGES_REQUESTED** [review]
Thank you for your contribution :)

I only have one small difference from the spec that should be fixed. Also we like to have the steps from the spec written as comments in the code (when it makes sense). For example:

```rust
// 1. Let O be ? ToObject(this value).
let obj = this.to_object(context)?;
```

It would be great if you could add those.

File: boa_engine/src/builtins/object/mod.rs (hunk 1 above)
Comment (CYBAI) [reply boa_engine/src/builtins/object/mod.rs]: > This should test if getter is callable not this. From the spec: If IsCallable(getter) is false.

oh! thank you so much for catching it! I now figured out why some tests doesn't pass 🙏 

> Also we can avoid the closure by using getter.is_callable() and then return context.throw_type_error() to throw the error.

Thank you for the tip! I've updated the commit.

**Review by raskad — APPROVED** [review]
Looks great! All tests in the `built-ins/Object/prototype/__defineGetter__` suite pass :)

**Review by jedel1043 — APPROVED** [review]
Looks good!

**Comment by CYBAI** [comment]
Thank you for the review! 🙏

**Pull request closed without merging**